File: Leetcode 75/Day 4 - Linked List/2. Linked List Cycle II (142)/LinkedListCycleII1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Naive Solution

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:

        hashSet = set()

        pointerNode = head

        while pointerNode and pointerNode.next:
            if pointerNode in hashSet:
                return pointerNode

            hashSet.add(pointerNode)
            pointerNode = pointerNode.next

        return None

File: Leetcode 75/Day 4 - Linked List/2. Linked List Cycle II (142)/LinkedListCycleII1.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */

public class Solution {
    public ListNode detectCycle(ListNode head) {

        Set<ListNode> hashSet = new HashSet<>();
        ListNode pointerNode = head;

        // Check each node to see if node is in the hash set
        // If it is, we just return the node
        // If it is not, we add it to the hash set

        while (pointerNode != null && pointerNode.next != null) {
            if (hashSet.contains(pointerNode)) {
                return pointerNode;
            } else {
                hashSet.add(pointerNode);
            }
            pointerNode = pointerNode.next;
        }
        return null;
    }
}